### PR TITLE
feat(statetree): implement optimized state tree diffing

### PR DESCRIPTION
**Pull request by frrist**
optimizes StateChangedActors by using the hamt diffign logic falling back to the lotus method when required, see code comments for specifics.

### CODE REVIEW

File: chain/indexer.go [hunk 1]
```diff
@@ -253,7 +258,7 @@ func (t *TipSetIndexer) TipSet(ctx context.Context, ts *types.TipSet) error {
 
 			// If we have actor processors then find actors that have changed state
 			if len(t.actorProcessors) > 0 {
-				changes, err := t.node.StateChangedActors(tctx, parent.ParentState(), child.ParentState())
+				changes, err := t.stateChangedActors(tctx, parent.ParentState(), child.ParentState())
```
Comment (frrist) [inline chain/indexer.go:261]: From the tracing added in subsequent PR's I've observed this optimized call reducing the runtime of the method from ~17sec to ~3sec.

File: chain/indexer.go [hunk 2]
```diff
@@ -408,6 +413,81 @@ func (t *TipSetIndexer) runProcessor(ctx context.Context, p TipSetProcessor, nam
...
+		if span.IsRecording() {
+			span.SetAttribute("diff", "fast")
+		}
+		changes, err := hamt.Diff(ctx, t.node.Store(), t.node.Store(), oldRoot, newRoot, hamt.UseTreeBitWidth(5), hamt.UseHashFunction(func(input []byte) []byte {
+			res := sha256.Sum256(input)
+			return res[:]
+		}))
+		if err != nil {
+			log.Errorw("failed to diff state tree efficiently, falling back to slow method", "error", err)
+		} else {
```
Comment (frrist) [inline chain/indexer.go:450]: In hindsight, implementing a state tree shim and using here could make this less :spaghetti:.

File: chain/indexer.go (hunk 1 above)
Comment (iand) [reply chain/indexer.go:261]: Very nice

**Review by iand — APPROVED** [review]

File: chain/indexer.go (hunk 2 above)
Comment (frrist) [reply chain/indexer.go:450]: Probably a day or two worth of effort (including review). I would start by extracting a modified version of the [lotus StateTree](https://github.com/filecoin-project/lotus/blob/master/chain/state/statetree.go#L159) to visor and implement some code gen, similar to the actor-shims, to keep it up to date with the specs-actors actor-states version.

**Pull request merged**